pkg/agent/externalnode/external_node_controller_test.go

```diff
@@ -0,0 +1,324 @@
+		},
+	}
+	portConfig := &interfacestore.OVSPortConfig{PortUUID: portUUID, OFPort: 3}
+	expectInterfaceConfig := &interfacestore.InterfaceConfig{
+		InterfaceName: "if1",
+		Type:          interfacestore.ExternalEntityInterface,
+		OVSPortConfig: &interfacestore.OVSPortConfig{PortUUID: portUUID, OFPort: 3},
+		EntityInterfaceConfig: &interfacestore.EntityInterfaceConfig{
+			EntityName:      "externalEntity",
+			EntityNamespace: "externalEntityNamespace",
+			UplinkPort: &interfacestore.OVSPortConfig{
+				PortUUID: randomUUID,
+				OFPort:   5,
+			},
+		},
+		IPs: []net.IP{
+			net.ParseIP("10.22.33.44"),
+			net.ParseIP("22.33.44.55"),
+		},
+	}
+
+	mockOVSBridgeClient.EXPECT().GetPortData(randomUUID, "uplinkName").Return(&ovsconfig.OVSPortData{OFPort: 5}, nil)
+	config, err := ParseHostInterfaceConfig(mockOVSBridgeClient, portData, portConfig)
+	assert.Equal(t, expectInterfaceConfig, config)
+	assert.NoError(t, err)
+}
+
+func TestReconcileHostUplinkFlows(t *testing.T) {
+	c := newExternalNodeController(t)
+
+	mockIfaceStore.EXPECT().GetInterfacesByType(interfacestore.ExternalEntityInterface).Return([]*interfacestore.InterfaceConfig{
+		&if1,
+		&if2,
+	})
+	mockOFClient.EXPECT().InstallVMUplinkFlows(if1.InterfaceName, int32(1), int32(5))
+	mockOFClient.EXPECT().InstallVMUplinkFlows(if2.InterfaceName, int32(2), int32(6))
+	c.reconcileHostUplinkFlows()
+}
+
+func TestEnqueueExternalNodeUpdate(t *testing.T) {
+	c := newExternalNodeController(t)
+	externalNode1 := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
+	externalNode2 := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm2", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
+	externalNode3 := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm3", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.3"}}},
+		},
+	}
+	testcases := []struct {
+		name           string
+		oldObj         *v1alpha1.ExternalNode
+		newObj         *v1alpha1.ExternalNode
+		expectEnqueued bool
+	}{
+		{
+			name:           "different externalnode",
+			newObj:         externalNode2,
+			oldObj:         externalNode1,
+			expectEnqueued: false,
+		},
+		{
+			name:           "same externalnode",
+			newObj:         externalNode3,
+			oldObj:         externalNode1,
+			expectEnqueued: true,
+		},
+	}
+	for _, tt := range testcases {
+		c.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "externalNode")
+		c.enqueueExternalNodeUpdate(tt.oldObj, tt.newObj)
+		if tt.expectEnqueued {
```

Comment:
Add "else case" to ensure the queue.Len() == 0 when no item is expected to enter the queue?